prov/shm/src/smr_ep.c

```diff
@@ -0,0 +1,747 @@
+{
+	ep->util_ep.rx_cq->src[ofi_cirque_windex(ep->util_ep.rx_cq->cirq)] =
+		(uint32_t) (uintptr_t) addr;
+	smr_rx_comp(ep, context, flags, len, buf, addr);
+}
+
+static void smr_rx_comp_signal(struct smr_ep *ep, void *context,
+			uint64_t flags, size_t len, void *buf, void *addr)
+{
+	smr_rx_comp(ep, context, flags, len, buf, addr);
+	ep->util_ep.rx_cq->wait->signal(ep->util_ep.rx_cq->wait);
+}
+
+static void smr_rx_src_comp_signal(struct smr_ep *ep, void *context,
+			uint64_t flags, size_t len, void *buf, void *addr)
+{
+	smr_rx_src_comp(ep, context, flags, len, buf, addr);
+	ep->util_ep.rx_cq->wait->signal(ep->util_ep.rx_cq->wait);
+
+}
+
+void smr_progress_resp(struct smr_ep *ep)
+{
+	struct smr_resp *resp;
+
+	fastlock_acquire(&ep->region->lock);
+	while (!ofi_cirque_isempty(smr_resp_queue(ep->region))) {
+		resp = ofi_cirque_head(smr_resp_queue(ep->region));
+		if (resp->status)
+			break;
+		ep->tx_comp(ep, NULL);
+		ofi_cirque_discard(smr_resp_queue(ep->region));
+	}
+	fastlock_release(&ep->region->lock);
+}
+
+void smr_progress_cmd(struct smr_ep *ep)
+{
+	struct smr_ep_entry *entry;
+	struct smr_cmd *cmd;
+	struct smr_resp *resp;
+	struct smr_inject_buf *tx_buf;
+	size_t inj_offset;
+	struct smr_region *peer_smr;
+	fi_addr_t addr;
+	int peer_id, ret = 0;
+
+	fastlock_acquire(&ep->region->lock);
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
+	while (!ofi_cirque_isempty(smr_cmd_queue(ep->region))) {
+		if (ofi_cirque_isempty(ep->rxq))
+			break;
+
+		entry = ofi_cirque_head(ep->rxq);
+		cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
+
+		switch (cmd->hdr.op.op_data){
+		case smr_op_inline:
+			ret = ofi_copy_to_iov(entry->iov, entry->iov_count, 0,
+					cmd->data.msg, cmd->hdr.op.size);
+			if (ret != cmd->hdr.op.size) {
+				FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+					"recv truncated");
+				ret = -FI_EIO;
+			} else {
+				ret = 0;
+			}
+			break;
+		case smr_op_inject:
+			inj_offset = (size_t) cmd->hdr.op.data;
+			tx_buf = (struct smr_inject_buf *) ((char **) ep->region +
+							    inj_offset);
+			ret = ofi_copy_to_iov(entry->iov, entry->iov_count, 0,
+					tx_buf, cmd->hdr.op.size);
+			if (ret != cmd->hdr.op.size) {
+				FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+					"recv truncated");
+				ret = -FI_EIO;
+			} else {
+				ret = 0;
+			}
+			freestack_push(smr_inject_pool(ep->region), tx_buf);
+			break;
+		case smr_op_iov:
+			peer_id = *(int *)ofi_av_get_addr(ep->util_ep.av,
+							  entry->addr);
+			peer_smr = smr_peer_region(ep->region, peer_id);
+			resp = (struct smr_resp *) ((char **) peer_smr +
+						    (size_t) cmd->hdr.op.data);
+			resp->cmd_id = cmd->hdr.cmd_id;
+
+			#if !defined __APPLE__ && !defined _WIN32
+				ret = process_vm_readv(peer_smr->pid, entry->iov,
+					       entry->iov_count, cmd->data.iov,
+					       cmd->hdr.op.iov_count, 0);
+			#else
+				FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+					"cannot transfer between processes\n");
+				ret = -1;
+				errno = FI_ENOSYS;
+			#endif
```

Comment:
This needs to be part of a compile check to disable the shm provider on those systems.  (Pending an implementation of some other transfer mechanism that will work.)